### PR TITLE
#159082372 Users receive a JWT on successful Registration

### PR DESCRIPTION
**Pull request by lytes20**
#### What does this PR do?
- Users should receive a JWT on successful Registration
#### Description of Task to be completed?
- Users once registered via the registration endpoint, should receive a JWT to be required on all subsequent calls to all other endpoints that require authentication.
#### How should this be manually tested?
- Running the endpoint in postman
#### What are the relevant pivotal tracker stories?
[#159082372](https://www.pivotaltracker.com/story/show/159082372)
#### Screenshots
screen shot 2018-07-31 at 19 11 59

#### Questions:
Am currently blocked by raising the coverage of the tests, anyway on how i can do this would be welcome

### CODE REVIEW

File: .coveragerc [hunk 1]
```diff
@@ -10,6 +10,7 @@ omit =
     *manage*
     *0001_initial*
     */serializers.py*
+    */backends.py*
```
Comment (geofrocker) [inline .coveragerc]: Why did you eliminate this

Comment (lytes20) [reply .coveragerc]: I eliminated this because I thought it wasn't relevant to test this as it is more of like a validation file

Comment (geofrocker) [reply .coveragerc]: What happens when you don't ignore it

Comment (lytes20) [reply .coveragerc]: It reduces coverage

**Review by abulojoshua1 — APPROVED** [review]

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -9,16 +9,61 @@
...
+    def _authenticate_credentials(self, request, token):
         """
-        Check Authenticity of a user token here
+        Method to try and authenticate the given credentials.
         """
-        pass
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+        except:  # noqa: E722
+            msg = 'Invalid authentication. Could not decode token.'
```
Comment (abulojoshua1) [inline authors/apps/authentication/backends.py:53]: @lytes20, please change this message to a more user-friendly message. For example `please log in to access this page`.

File: authors/apps/authentication/backends.py [hunk 3]
```diff
@@ -9,16 +9,61 @@
...
+            raise exceptions.AuthenticationFailed(msg)
+
+        try:
+            user = User.objects.get(pk=payload['id'])
+        except User.DoesNotExist:
+            msg = 'No user matching this token was found.'
+            raise exceptions.AuthenticationFailed(msg)
+
+        if not user.is_active:
+            msg = 'This user has been deactivated.'
```
Comment (abulojoshua1) [inline authors/apps/authentication/backends.py:63]: I would recommend using a message like `msg = please, check your email to activate your account`. This is because a user should only be deactivated after he/she deletes his/her account and not when his/her account is not yet verified by email. In the modal.py file, the is a database row created named `is_active`. Check it out and read the documentation above that line.

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (lytes20) [reply authors/apps/authentication/backends.py:53]: Lol, so this was your task @abulojoshua1  to write more descriptive and meaningful messages but will do it. thanks

File: authors/apps/authentication/backends.py (hunk 3 above)
Comment (lytes20) [reply authors/apps/authentication/backends.py:63]: Yes, so if we're to change this message, I think we could change it to `this account has been deactivated`. because we are not providing a reactivation feature for now. Users, I think, will just have to create a new account once they deactivated/deleted their account.

Comment (abulojoshua1) [reply authors/apps/authentication/backends.py:63]: @lytes20, There is a difference between a deactivated account and the situation where one has not confirmed their email address. For the case in the subject, the user has not yet approved their email. `is_active` in models.py is by default equal to boolean `True`. This means that the user's account is not deleted and he/she can still use the account. But, `is_account_verfied` in your models.py file is by default set to boolean `False`, meaning that the user's account is not yet verified as they have not clicked the link in their email address and so they cannot log in.

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (abulojoshua1) [reply authors/apps/authentication/backends.py:53]: @lyte, I do understand what you mean there, but I was writing descriptive messages for already existing functionality and not new functionality. Currently, your branch has not been merged to develop and so I do not have your changes to modify that line. Most of my modification were in `test_views.py` and `serializers.py` files not `backend.py` file.

File: authors/apps/authentication/backends.py (hunk 3 above)
Comment (abulojoshua1) [reply authors/apps/authentication/backends.py:63]: My bad, @lytes20. Sorry, I misunderstood that line of code. You are actually right on that, the code acts as is supposed to.

File: authors/apps/authentication/backends.py (hunk 2 above)
Comment (geofrocker) [reply authors/apps/authentication/backends.py:53]: @lytes20 Is this clear to you

File: authors/apps/authentication/serializers.py [hunk 4]
```diff
@@ -133,7 +140,7 @@ def validate(self, data):
         return {
             'email': user.email,
             'username': user.username,
-
```
Comment (ssewilliam) [inline authors/apps/authentication/serializers.py:136]: Why are we making a new line here

File: authors/apps/authentication/tests/test_jwt.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        """
+        Reusable function to login a user
+        """
+
+        response = self.test_client.post(
+            "/api/users/login/", data=json.dumps(self.data), content_type='application/json')
+        token = response.json()['user']['token']
+        return token
+
+        # return (headers = {'HTTP_AUTHORIZATION': 'Token ' + token })
```
Comment (ssewilliam) [inline authors/apps/authentication/tests/test_jwt.py]: @lytes20 
- We should remove uncommented code

File: authors/apps/authentication/tests/test_jwt.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        }
+
+        response = self.test_client.get(
+            "/api/user/", **headers, content_type='application/json')
+        print(response)
+        self.assertEqual(response.status_code, 401)
+
+    def test_token_passed_but_user_doesnt_exist(self):
+        """
+        Test to check when a wrong prefix is passed in the header
```
Comment (ssewilliam) [inline authors/apps/authentication/tests/test_jwt.py]: @lytes20 
- Shouldn't this test to check if token passed but user does not exist

File: authors/apps/authentication/tests/test_jwt.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        headers = {
+            'HTTP_AUTHORIZATION': 'Bear ' + self.token
+        }
+
+        response = self.test_client.get(
+            "/api/user/", **headers, content_type='application/json')
+        print(response)
+        self.assertEqual(response.status_code, 401)
+
+    def test_token_passed_but_user_doesnt_exist(self):
```
Comment (ssewilliam) [inline authors/apps/authentication/tests/test_jwt.py:136]: @lytes20 
- This test actually seems to do what the one on line 80 is doing
- Testing for a valid token

File: authors/apps/authentication/tests/test_models.py [hunk 8]
```diff
@@ -1,5 +1,8 @@
 from django.test import TestCase
 from ..models import User, UserManager
+import jwt
+from django.conf import settings
+from datetime import datetime, timedelta
```
Comment (ssewilliam) [inline authors/apps/authentication/tests/test_models.py:6]: @lytes20 
- I would recommend importing **python modules**, **first then Django**, *then your own**

File: authors/apps/authentication/tests/test_views.py [hunk 9]
```diff
@@ -1,6 +1,7 @@
 from django.test import Client
 from django.test import TestCase
 import json
+from ..models import User
```
Comment (ssewilliam) [inline authors/apps/authentication/tests/test_views.py:4]: @lytes20 
- I would recommend importing **python modules**, **first then Django**, *then your own**

File: authors/apps/authentication/views.py [hunk 10]
```diff
@@ -49,7 +49,7 @@ def post(self, request):
 
 
 class UserRetrieveUpdateAPIView(RetrieveUpdateAPIView):
-    permission_classes = (IsAuthenticated,)
+    permission_classes = (IsAuthenticated, )
```
Comment (ssewilliam) [inline authors/apps/authentication/views.py]: @lytes20 
- Was this change necessary?

**Review by ssewilliam — COMMENTED** [review]
@lytes20 
I would recommend allowing more reusability of code and checking a few typos...

Otherwise this work is **Super Impressive**

File: authors/apps/authentication/tests/test_jwt.py (hunk 5 above)
Comment (lytes20) [reply authors/apps/authentication/tests/test_jwt.py]: yes, yes, let me work on this now

File: authors/apps/authentication/tests/test_jwt.py (hunk 6 above)
Comment (lytes20) [reply authors/apps/authentication/tests/test_jwt.py]: gonna change the comment, thanks

File: authors/apps/authentication/serializers.py (hunk 4 above)
Comment (lytes20) [reply authors/apps/authentication/serializers.py:136]: I actually just deleted the line for neatness

Comment (abulojoshua1) [reply authors/apps/authentication/serializers.py:136]: @ssewilliam, the line was just removed and not added.

File: authors/apps/authentication/tests/test_jwt.py (hunk 7 above)
Comment (lytes20) [reply authors/apps/authentication/tests/test_jwt.py:136]: no not really @ssewilliam , so there is a validation for an incidence when a user might generate a token that actually has the right signature but is not of a user in our database hence the test for this as well

File: authors/apps/authentication/views.py (hunk 10 above)
Comment (abulojoshua1) [reply authors/apps/authentication/views.py]: @ssewilliam, That was probably caused by him performing PEP8 formating.

File: authors/apps/authentication/tests/test_models.py (hunk 8 above)
Comment (lytes20) [reply authors/apps/authentication/tests/test_models.py:6]: cool, will start doing this moving forward

File: authors/apps/authentication/views.py (hunk 10 above)
Comment (lytes20) [reply authors/apps/authentication/views.py]: no, it wasn't, correcting this now

**Review by ssewilliam — APPROVED** [review]
Impressive test coverage
